src/compiler/visitorPublic.ts

```diff
@@ -159,29 +212,29 @@ namespace ts {
 
         // Visit each original node.
         for (let i = 0; i < count; i++) {
-            const node: T = nodes[i + start];
-            const visited = node !== undefined ? visitor(node) : undefined;
-            if (updated !== undefined || visited === undefined || visited !== node) {
+            const node = nodes[i + start];
+            const visited = node !== undefined ? (visitor ? visitor(node) : node) : undefined;
+            if (updated !== undefined || visited === undefined || visited !== node as VisitResult<Node>) {
                 if (updated === undefined) {
                     // Ensure we have a copy of `nodes`, up to the current index.
                     updated = nodes.slice(0, i);
                 }
                 if (visited) {
                     if (isArray(visited)) {
                         for (const visitedNode of visited) {
-                            void Debug.assertNode(visitedNode, test);
-                            updated.push(visitedNode as T);
+                            updated.push(visitedNode);
                         }
                     }
                     else {
-                        void Debug.assertNode(visited, test);

```

Comment:
Probably, it'd be best to check this here too, for instantish feedback, but the later assert still hast to run in case the nodes are not updated.

That, or we check `node` instead if the pushed doesn't happen.